src/adventure/adventuresession.cpp

```diff
@@ -0,0 +1,108 @@
+#include "adventuresession.h"
+#include "QtCore/qdebug.h"
+#include "adventuretracker.h"
+#include "adventurewidget.h"
+#include "xpstatuswidget.h"
+
+AdventureSession::AdventureSession(QString charName)
+    : m_charName{charName}
+    , m_startTimePoint{std::chrono::steady_clock::now()}
+    , m_endTimePoint{}
+    , m_isEnded{false}
+    , m_xpInitial{0.0}
+    , m_xpCheckpoint{0.0}
+    , m_xpCurrent{0.0}
+{}
+
+AdventureSession::AdventureSession(const AdventureSession &src)
+    : m_charName{src.m_charName}
+    , m_startTimePoint{src.m_startTimePoint}
+    , m_endTimePoint{src.m_endTimePoint}
+    , m_isEnded{src.m_isEnded}
+    , m_xpInitial{src.m_xpInitial}
+    , m_xpCheckpoint{src.m_xpCheckpoint}
+    , m_xpCurrent{src.m_xpCurrent}
+{}
+
+AdventureSession &AdventureSession::operator=(const AdventureSession &src)
+{
+    m_charName = src.m_charName;
+    m_startTimePoint = src.m_startTimePoint;
+    m_endTimePoint = src.m_endTimePoint;
+    m_isEnded = src.m_isEnded;
+    m_xpInitial = src.m_xpInitial;
+    m_xpCheckpoint = src.m_xpCheckpoint;
+    m_xpCurrent = src.m_xpCurrent;
+
+    return *this;
+}
+
+void AdventureSession::endSession()
+{
+    m_isEnded = true;
```

Comment:
Should we add an assertion to avoid ending a session twice?

Reply:
Thanks, will do

Reply:
fixed